Abins: Use Euphonic Castep parser

ajjackson opened the pull request:
**Description of work.**

Abins reads data from ab-initio vibration/phonon calculations in a variety of formats, including .phonon files from CASTEP.

The inbuilt CASTEP parser makes the unsafe assumption that Gamma-points with LO-TO splitting, which use a longer header line to contain extra information, will always appear first. If this point appears later in the data, earlier k-points are split.

This assumption is quite deeply baked into the logic of the parser, requiring refactoring to address. As Mantid now depends on Euphonic, which contains a .phonon file parser /without/ this defect, this work would be redundant.

Here we remove the inbuilt parser for CASTEP .phonon files and delegate this to the Euphonic library implementation.

We also bump the Euphonic version requirement and drop the importlib_resources pin (see #35798)

**To test:**
Unit tests and systemtests have been updated.

Fixes #33755 
Fixes #35798 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

ajjackson committed:
Abins: Use Euphonic Castep parser where available

The inbuilt Castep parser makes the unsafe assumption that
Gamma-points with LO-TO splitting, which use a longer header line to
contain extra information, will always appear first. If this point
appears later in the data, earlier k-points are split.

This assumption is quite deeply baked into the logic of the parser,
requiring refactoring to address. As we intend in any case to
increasingly depend on Euphonic, which contains a .phonon file parser
/without/ this defect, this work would be redundant.

Here we
- detect the unsafe behaviour in the inbuilt parser and raise an error
- use Euphonic if possible

Not all Mantid builds include Euphonic yet, but when we move to
Conda-based packaging for all builds the legacy implementation can be
removed.

Abins is structured around the idea that each format has a class
implementing the loader, so I have created a wrapper class around the
two implementations. This will make it very easy to remove the old
implementation in future. For now, the use_euphonic flag can be used
in testing to force one implementation or the other.

ajjackson committed:
Abins CastepLoader changes: simpler class setup, legacy-only tests

- Having different classes and wrapping them in a shim layer was
  proving a bit of a headache. Dispatching to different methods seems
  simpler and looks the same to existing API.

- Allow use_euphonic flag to be passed to test suite, so different
  test data can be used with each implementation. (We expect different
  results, and at this stage want to keep track of when they
  change...)
  - A change in legacy results could indicate a mistake in this
    refactor/shim process.
  - A change in Euphonic results could indicate an upstream change;
    again this is probably a good thing but we should keep track of it
    for now.

- Maybe when things have settled down, Euphonic is trusted/stable and
  ubiquitous, then all these tests can be replaced by mocking!

ajjackson committed:
Abins Castep import: update systemtest

This is an unfortunate case: the results are expected to depend on the
installed libraries, because the inbuilt and Euphonic loaders handle
some sample files differently. (This is due to the ambiguous weighting
of LO-TO-split Gamma point data in .phonon files.)

As the systemtest cannot change the low-level arguments to force one
backend or the other, and could plausibly be run on either system,
we provide test data for both cases.

When Conda migration is complete, we should be able to remove the
legacy loader and simplify things.

ajjackson committed:
Abins tests: trouble finding some files

Gaussian files are being found successfully on my (Linux+Conda)
machine and Windows CI but not other CI tests. Suspect some
inconsistency in case-sensitivity handling; tweak so that empty
return values are treated same as error.

ajjackson committed:
Abins Castep/Euphonic import: simple mocking test

I'm not 100% happy with this as the "except" could catch a lot of
different things. Better would perhaps be to mock a return value
for QpointPhononModes.from_castep, then we can thoroughly check
the code in Abins and _only_ in Abins.

But then we would be entirely reliant on the systemtest to catch
upstream changes. Perhaps that's good: upstream changes are then
caught in only one or two places, so much easier to respond to with
updated data.

Anyway, to improve this further we really need better test coverage of
the EuphonicLoader and its data reformatting methods. Which might be a
separate bit of work, and coupled to caching?

ajjackson committed:
Abins: Euphonic-Castep import - release note

ajjackson committed:
Castep/Euphonic test: skipUnless does nothing, try moving it

Maybe it only works on individual test methods?

AnthonyLim23 commented:
do we want to update this now we have moved to conda?

ajjackson commented:
If we are 100% Conda builds now, then we could remove the legacy parser entirely which would be nice.

As mentioned, work is ongoing on the Euphonic parser. Most likely those changes will be part of Euphonic 1.x along with some API changes. (I pinned the Conda dependency to 0.6.x for a reason...) So it could make sense to merge this before beginning an "update to Euphonic 1.x (w/ API changes)" PR once that is ready.

ajjackson commented:
Might close later but not stale today

ajjackson committed:
Black demands tribute: mostly whitespace

ajjackson commented:
As far as I can tell there are no merge conflicts and it is simply unable to remove the label for permissions reasons. We've seen this before.

ajjackson committed:
Euphonic is now a required dependency: remove checks and legacy code

- Abins CASTEP importer can now depend solely on Euphonic
- A bit of logic is also added to handle API break in version 1.0 so
  we can stagger the update of code and dependency
- Checks on Euphonic availability can be removed in various places

ajjackson committed:
Abins test data update: invalid CASTEP .phonon file

The squaricn_no_sum_LoadCASTEP test case uses a hand-tweaked file.
This file inadvertantly includes an extra blank line, which is
accepted by the Abins native .phonon file loader but not the external
Euphonic loader. As it is technically invalid, we can update the test
data rather than change the robustness of Euphonic loader.

(It might be an idea to do that anyway, but it need not hold up
refactoring work.)

ajjackson committed:
Remove remaining calls to (removed) euphonic_available function

ajjackson commented:
The change to the *squareicn_no_sum_LoadCASTEP* test data is very subtle; only a trailing newline was changed (and, as a result, the expected checksum is completely different). This is because the handcrafted file is not valid input to the Euphonic reader (as it is slightly off-spec) but was tolerated by the native reader being removed in this PR.

ajjackson commented:
Note that after rebasing onto #35532 (in its current state), the unit tests related to powder calculations are now passing - this branch decouples those tests from CASTEP parsing. The CASTEP loader test still has a couple of failures to address, but this is expected with the implementation change.

ajjackson committed:
Update Euphonic conda requirement

ajjackson committed:
Castep loader: prefer_non_loto and strip zero-weighted points

This feature was added in Euphonic 1.2.1 and gives more appropriate
behaviour when LO-TO splitting is included in the .phonon file but
data is also available at a non-split point.

This is implemented in Euphonic by setting the weights of undesired
points to zero: for simulation purposes we can strip these out
entirely and save some calculation effort.

ajjackson commented:
Running this locally I get a couple of test failures, both of which are explainable:

- some atomic masses do not match the reference data. This is because Euphonic reads the values from the file while the legacy Mantid parser does some tweaking to match them to nearby values from Mantid. It's not clear that this is necessary and it makes testing more confusing.

- The legacy parser chooses the wrong set of Gamma-point data from squaricn_sum_LoadCASTEP.phonon in "test_gamma_sum_correction". This is discussed in the (incorrectly closed as stale) parent Issue https://github.com/mantidproject/mantid/issues/33755
  - In this file, the Gamma point appears twice: first with a short line for "exact" data at **q**=[0, 0, 0], then another set of data is given with a longer specification line indicating a small perturbation away from **q**=[0, 0, 0].
  - The existing parser skips the first of those points and favours the second. Scientifically this is a bad choice; as discussed in Euphonic Issues such as https://github.com/pace-neutrons/Euphonic/issues/273 we get better statistics with the exact [0, 0, 0] point.
  - (Ideally users would not include the Gamma-point at all when sampling systems with LO-TO splitting, but this is beyond our control.)

ajjackson committed:
Remove unused properties from CASTEPLoader

ajjackson committed:
Update CASTEPLoader test data

For the case "squaricn_sum_LoadCASTEP", existing test cases expects us
to only use the data for [0, 0, 0] -> [1, 0, 0].

This is a bad choice, and with the "prefer_non_loto" argument the
Euphonic-based parser instead takes the (exact) [0, 0, 0] data.

ajjackson committed:
Temporarily relax Abins loader mass check

To confirm that this difference in the masses read is the only cause
of test failures at this point.

ajjackson committed:
Add a small absolute tolerance to coordinates comparisons

Abins Testing checks positions against reference files. We are seeing
a failure on Windows of around 1.1e-16; i.e. around the floating-point
precision... This could be coming from the unit conversions in
Pint/Euphonic. Set a (still absurdly small) tolerance for this.

ajjackson commented:
Ah I see, Mantid is currently pinned to LLVM version 15

ajjackson committed:
Revert mass comparison delta, update isotope test data

Update the reference data for isotope masses test.  This change
shouldn't have any impact on downstream Abins calculations; masses are
rounded to the nearest integer for purposes of determining the
cross-section!

ajjackson commented:
> 17:39:56 Encountered problems while solving:
> 17:39:56   - package euphonic-1.2.1-py310h4148280_0 requires llvm-openmp >=16.0.6, but none of the providers can be installed

grr, Mac version is still using the wrong Conda build. We shouldn't be pinning exact builds, any idea how to kick it to use the version 1 build which requires openmp >= 15?

ajjackson commented:
Linux failure is system tests, that's probably to be expected but I'll need to look more closely and update the data if so.

The Mac situation is weird, installing from mantid-developer-osx.yml looks for the wrong build and fails.

Installing euphonic=1.2.1=py38hd2faf92_1 works.

If I _then_ try to install from the yaml file it tries to install a Python 3.9-based euphonic build and gives up.

ajjackson commented:
Somehow the Mac Conda build is requiring llvm-opemp version >=15 *and* >=16, even though only 15 is asked for explicitly.

I have been chatting with @jaimergp of conda-forge on their Element channel; he has been very helpful and suggested a couple of things to try.

ajjackson commented:
OSX passing, woohoo :tada:

ajjackson committed:
Update systemtest using CASTEP importer

ajjackson committed:
Docstring updates

ajjackson committed:
Fix Euphonic version check

Not clear how this wasn't breaking before? Anyway, input to
version.parse is supposed to be a string.

ajjackson committed:
Update release note

jmborr commented on scripts/abins/atomsdata.py:
```diff
@@ -92,15 +92,23 @@ def _check_item(item: _AtomData, n_atoms: Optional[int] = None) -> _AtomData:
+                symbol = symbol_raw.decode("utf-8")
+            else:
+                symbol = symbol_raw
+                assert isinstance(symbol, str)
```
perhaps a bit more informative error?
```python
elif isinstance(symbol, str):
    symbol = symbol_raw
else:
    raise TypeError(f"Invalid type for symbol {symbol_raw}")
```

ajjackson replied:
This `assert` is for mypy's benefit; it seems to be the only way to convince it that a variable that could once hold more than one type now holds a specific type.

I'm not sure that will work if it isn't a fairly simple `assert` statement.

jmborr commented on scripts/abins/input/tester.py:
```diff
@@ -206,21 +215,30 @@ def check(self, *, name: str, loader: AbInitioLoader, extension: str = None, max
         except ValueError:
+            read_filename = ""
+        if not read_filename:
             read_filename = abins.test_helpers.find_file(filename=filename + "." + extension.upper())
```
if a `ValueError` was procured to deal with line:
```python
read_filename = abins.test_helpers.find_file(filename=filename + "." + extension)
```
then perhaps one could procure another `ValueError` to deal with line
```python
read_filename = abins.test_helpers.find_file(filename=filename + "." + extension.upper())
```
?

ajjackson replied:
I think the goal here was to avoid a confusing nested try-except, and always hit the same error call. But I the variable usage here is also a bit confusing.

This pattern is a bit ugly either way and appears in a few places in this file. Part of the answer might be to make extension-case-insensitivity a feature of `abins.test_helpers.find_file`; we can actually dodge all the try-except logic by looking at the filename before passing it to to the load function.

ajjackson replied:
Ok, that looks better to me!

ajjackson committed:
Refactor file-finder in Abins tests

This is already delegated to a function, let that take responsibility
for missing files and case-sensitivity.

ajjackson commented:
Funny build errors, will try rebasing?

jmborr commented on conda/recipes/conda_build_config.yaml:
```diff
@@ -79,10 +79,7 @@ readline:
-
-importlib_resources:
-  - 5.12.0
+  - '>=1.2.1 <2.0'
```
I think you need `,`
```bash
'>=1.2.1,<2.0'
```
https://docs.conda.io/projects/conda-build/en/stable/resources/package-spec.html#package-match-specifications

ajjackson committed:
Conda build config: range formatting

Spotted by @jmborr

ajjackson commented:
Hmm, still getting funny build errors like
```
15:16:20 [1/909] Generating workbench resources module
15:16:20 FAILED: /jenkins_workdir/workspace/pull_requests-conda-linux/qt/applications/workbench/workbench/app/resources.py 
15:16:20 cd /jenkins_workdir/workspace/pull_requests-conda-linux/build/qt/applications/workbench && /jenkins_workdir/workspace/pull_requests-conda-linux/mambaforge/envs/mantid-developer/bin/python3.10 -m PyQt5.pyrcc_main -o /jenkins_workdir/workspace/pull_requests-conda-linux/qt/applications/workbench/workbench/app/resources.py /jenkins_workdir/workspace/pull_requests-conda-linux/qt/applications/workbench/resources.qrc
15:16:20 /bin/sh: /jenkins_workdir/workspace/pull_requests-conda-linux/mambaforge/envs/mantid-developer/bin/python3.10: No such file or directory
15:16:20 [2/909] Generating scripts .pth file
15:16:20 [3/909] Building CXX object Framework/Parallel/CMakeFiles/EventParallelLoader.dir/src/IO/EventLoaderHelpers.cpp.o
15:16:20 [4/909] Generating mantidqt resources module
15:16:20 FAILED: /jenkins_workdir/workspace/pull_requests-conda-linux/qt/python/mantidqt/mantidqt/resources.py 
15:16:20 cd /jenkins_workdir/workspace/pull_requests-conda-linux/build/qt/python/mantidqt && /jenkins_workdir/workspace/pull_requests-conda-linux/mambaforge/envs/mantid-developer/bin/cmake -P create_resources.cmake
15:16:20 CMake Error at create_resources.cmake:5 (string):
15:16:20   string sub-command REPLACE requires at least four arguments.
15:16:20 
15:16:20 
```

MialLewis commented:
> > Hmm, still getting funny build errors like
> > ```
> > 15:16:20 [1/909] Generating workbench resources module
> > 15:16:20 FAILED: /jenkins_workdir/workspace/pull_requests-conda-linux/qt/applications/workbench/workbench/app/resources.py 
> > 15:16:20 cd /jenkins_workdir/workspace/pull_requests-conda-linux/build/qt/applications/workbench && /jenkins_workdir/workspace/pull_requests-conda-linux/mambaforge/envs/mantid-developer/bin/python3.10 -m PyQt5.pyrcc_main -o /jenkins_workdir/workspace/pull_requests-conda-linux/qt/applications/workbench/workbench/app/resources.py /jenkins_workdir/workspace/pull_requests-conda-linux/qt/applications/workbench/resources.qrc
> > 15:16:20 /bin/sh: /jenkins_workdir/workspace/pull_requests-conda-linux/mambaforge/envs/mantid-developer/bin/python3.10: No such file or directory
> > 15:16:20 [2/909] Generating scripts .pth file
> > 15:16:20 [3/909] Building CXX object Framework/Parallel/CMakeFiles/EventParallelLoader.dir/src/IO/EventLoaderHelpers.cpp.o
> > 15:16:20 [4/909] Generating mantidqt resources module
> > 15:16:20 FAILED: /jenkins_workdir/workspace/pull_requests-conda-linux/qt/python/mantidqt/mantidqt/resources.py 
> > 15:16:20 cd /jenkins_workdir/workspace/pull_requests-conda-linux/build/qt/python/mantidqt && /jenkins_workdir/workspace/pull_requests-conda-linux/mambaforge/envs/mantid-developer/bin/cmake -P create_resources.cmake
> > 15:16:20 CMake Error at create_resources.cmake:5 (string):
> > 15:16:20   string sub-command REPLACE requires at least four arguments.
> > 15:16:20 
> > 15:16:20 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The `python3.10: No such file or directory` is a bit worrying
> 
> yes, that's what I'm scratching my head about. How does variable [Python_EXECUTABLE](https://github.com/ajjackson/mantid/blob/abins-castep-import/qt/CMakeLists.txt#L11) becomes `3.10` 🤔

Not looked into this specific issue, but the `Python_EXECUTABLE` variable is set here (or in relative places for our other packages), to $PYTHON which I believe is the python interpreter specified in the active conda environment. https://github.com/mantidproject/mantid/blob/39b6213d434256151533c73d10c9c09a8fa39a99/conda/recipes/mantid/build.sh#L24

ajjackson committed:
move release note forward a few versions

peterfpeterson commented:
The build issue has to do with sphinx warnings. You need to fix your release notes
```
08:48:34 release/v6.8.0/indirect_geometry:4: WARNING: Bullet list ends without a blank line; unexpected unindent.
08:48:34 release/v6.8.0/indirect_geometry:6: ERROR: Unexpected indentation.
```

jmborr committed:
add blank line

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
itemize the entry and wrap lines

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

ajjackson commented:
Thanks for cleaning this up while I was on leave 😄

jmborr commented:
no problem! 😄
